libcudacxx/include/cuda/__iterator/offset_iterator.h

```diff
@@ -0,0 +1,494 @@
+class offset_iterator
+{
+private:
+  _Iter __iter_         = {};
+  _OffsetType __offset_ = {};
+
+public:
+  using iterator_type     = _Iter;
+  using iterator_concept  = _CUDA_VSTD::random_access_iterator_tag;
+  using iterator_category = _CUDA_VSTD::random_access_iterator_tag;
+  using value_type        = _CUDA_VSTD::iter_value_t<_Iter>;
+  using difference_type   = _CUDA_VSTD::iter_difference_t<_Iter>;
+
+  using __offset_value_t      = typename __offset_iterator_offset_value_type<_OffsetType>::type;
+  using __offset_difference_t = typename __offset_iterator_offset_difference_type<_OffsetType>::type;
+
+  //! Ensure that the user passes either something integer_like or an iterator to something interger_like
+  static_assert(_CUDA_VSTD::__integer_like<__offset_value_t>,
+                "cuda::offset_iterator: _OffsetType must either be interger_like or an iterator to integer_like");
+
+  //! Ensure that the offset type is convertible to difference_type
+  static_assert(_CUDA_VSTD::is_convertible_v<__offset_value_t, difference_type>,
+                "cuda::offset_iterator: _OffsetType must either be convertible to iter_difference<Iter> or an iterator "
+                "to a type convertible to iter_difference<Iter>");
+
+  //! To actually use operator+ we need the base iterator to be random access
+  static_assert(_CUDA_VSTD::random_access_iterator<_Iter>,
+                "cuda::offset_iterator: _Iter must be a random access iterator!");
+  //! To actually use operator+ we need the offset iterator to be random access
+  static_assert(_CUDA_VSTD::__integer_like<_OffsetType> || _CUDA_VSTD::random_access_iterator<_OffsetType>,
+                "cuda::offset_iterator: _OffsetType must either be an integer type or a random access iterator!");
+
+  //! @brief Default constructs an \p offset_iterator with a value initialized iterator and offset
+  _CCCL_EXEC_CHECK_DISABLE
+  _CCCL_HIDE_FROM_ABI constexpr offset_iterator() = default;
+
+  //! @brief Constructs an \p offset_iterator from an iterator and an optional offset
+  //! @param __iter The iterator to to offset from
+  //! @param __offset Optional offset. The default is a value initialized _OffsetType
+  _CCCL_EXEC_CHECK_DISABLE
+  _LIBCUDACXX_HIDE_FROM_ABI constexpr offset_iterator(_Iter __iter, _OffsetType __offset = {}) noexcept(
+    _CUDA_VSTD::is_nothrow_copy_constructible_v<_Iter> && _CUDA_VSTD::is_nothrow_copy_constructible_v<_OffsetType>)
+      : __iter_(__iter)
+      , __offset_(__offset)
+  {}
+
+  //! @brief Returns a const reference to the stored iterator we are offsetting from
+  _CCCL_EXEC_CHECK_DISABLE
+  [[nodiscard]] _LIBCUDACXX_HIDE_FROM_ABI constexpr const _Iter& base() const& noexcept
+  {
+    return __iter_;
+  }
+
+  //! @brief Extracts the stored iterator we are offsetting from
+  _CCCL_EXEC_CHECK_DISABLE
+  [[nodiscard]] _LIBCUDACXX_HIDE_FROM_ABI constexpr _Iter base() &&
+  {
+    return _CUDA_VSTD::move(__iter_);
+  }
```

Comment:
Why don't we offer mutable access to the base iterator? `_Iter& base()`?

Reply:
Because that might be unsafe. `ranges` try to be really explicit in what you can or cannot do. changing the underlying iterator under the hood is dangerous and could potentially break class invariants